File: go/raw-chat/indexer.go
package main

import (
	"bufio"
	"crypto/sha256"
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

// JsonlRecord represents a single JSONL line from a session file.
type JsonlRecord struct {
	Type      string `json:"type"`
	ID        string `json:"id"`
	Timestamp int64  `json:"timestamp"`
	Message   *struct {
		Role      string      `json:"role"`
		Content   interface{} `json:"content"`
		ID        string      `json:"id"`
		ParentID  string      `json:"parentId"`
		Timestamp int64       `json:"timestamp"`
	} `json:"message"`
}

// WatermarkRow represents the indexing progress for a single JSONL file.
type WatermarkRow struct {
	SourceFile      string
	SizeBytes       int64
	MtimeMs         int64
	LastLine        int
	LastIndexedAtMs int64
}

// IndexSessionFile indexes a session JSONL file incrementally.
// Uses watermarks to only process new lines.
// Returns counts of indexed, skipped, and errored records.
func IndexSessionFile(db *sql.DB, sourceFile string, agentID string, sessionKey string) (indexed, skipped, errors int) {
	stat, err := os.Stat(sourceFile)
	if err != nil {
		return 0, 0, 0
	}

	// Check watermark.
	wm := getWatermark(db, sourceFile)
	if wm != nil && wm.SizeBytes == stat.Size() && wm.MtimeMs == stat.ModTime().UnixMilli() {
		return 0, 0, 0 // File hasn't changed.
	}
	lastLine := 0
	if wm != nil {
		lastLine = wm.LastLine
	}

	// Read the file.
	f, err := os.Open(sourceFile)
	if err != nil {
		return 0, 0, 0
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 1024*1024), 4*1024*1024) // 4MB max line

	lineNum := 0
	var lines []string
	for scanner.Scan() {
		lineNum++
		if lineNum <= lastLine {
			continue // Skip already-indexed lines.
		}
		lines = append(lines, scanner.Text())
	}

	if len(lines) == 0 {
		return 0, 0, 0
	}

	now := time.Now().UnixMilli()

	// Use a transaction for atomicity.
	tx, err := db.Begin()
	if err != nil {
		return 0, 0, 0
	}
	defer tx.Rollback()

	insertMsg, err := tx.Prepare(`
		INSERT OR IGNORE INTO chat_messages (
			stable_key, session_id, session_key, agent_id, channel,
			message_id, parent_id, role, timestamp_ms, timestamp_iso,
			date_key, text, raw_json, source_file, source_line, indexed_at_ms
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`)
	if err != nil {
		return 0, 0, 0
	}
	defer insertMsg.Close()

	insertFts, err := tx.Prepare(`
		INSERT OR IGNORE INTO chat_messages_fts (rowid, text, role, session_id, timestamp_iso)
		SELECT id, text, role, session_id, timestamp_iso FROM chat_messages WHERE id = ?
	`)
	if err != nil {
		// FTS may not be available; continue without it.
		insertFts = nil
	}
	defer func() {
		if insertFts != nil {
			insertFts.Close()
		}
	}()

	for i, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}

		var record JsonlRecord
		if err := json.Unmarshal([]byte(line), &record); err != nil {
			errors++
			continue
		}

		// Only index message records.
		if record.Type != "message" || record.Message == nil {
			continue
		}

		msg := record.Message
		if msg.Role == "" {
			errors++
			continue
		}

		text := extractText(msg.Content)
		if text == "" {
			skipped++
			continue
		}

		stableKey := buildStableKey(record, sourceFile, lastLine+i+1, line)
		timestampMs := msg.Timestamp
		if timestampMs == 0 {
			timestampMs = record.Timestamp
		}
		if timestampMs == 0 {
			timestampMs = now
		}
		timestampIso := time.UnixMilli(timestampMs).UTC().Format(time.RFC3339)
		dateKey := timestampIso[:10]

		channel := extractChannelFromSessionKey(sessionKey)
		messageID := msg.ID
		parentID := msg.ParentID
		sessionID := extractSessionID(record, sourceFile)

		result, err := insertMsg.Exec(
			stableKey, sessionID, sessionKey, agentID, channel,
			messageID, parentID, msg.Role, timestampMs, timestampIso,
			dateKey, text, line, sourceFile, lastLine+i+1, now,
		)
		if err != nil {
			errors++
			continue
		}

		// Check if the row was actually inserted (not ignored due to duplicate).
		rowsAffected, _ := result.RowsAffected()
		if rowsAffected > 0 {
			indexed++
			// Insert into FTS if available.
			if insertFts != nil {
				lastID, _ := result.LastInsertId()
				if lastID > 0 {
					insertFts.Exec(lastID)
				}
			}
		} else {
			skipped++
		}
	}

	// Update watermark.
	updateWatermark(tx, sourceFile, stat.Size(), stat.ModTime().UnixMilli(), lastLine+len(lines), now)

	if err := tx.Commit(); err != nil {
		return 0, 0, 0
	}

	return indexed, skipped, errors
}

func getWatermark(db *sql.DB, sourceFile string) *WatermarkRow {
	var wm WatermarkRow
	err := db.QueryRow(
		`SELECT source_file, size_bytes, mtime_ms, last_line, last_indexed_at_ms FROM chat_index_watermarks WHERE source_file = ?`,
		sourceFile,
	).Scan(&wm.SourceFile, &wm.SizeBytes, &wm.MtimeMs, &wm.LastLine, &wm.LastIndexedAtMs)
	if err != nil {
		return nil
	}
	return &wm
}

func updateWatermark(tx *sql.Tx, sourceFile string, sizeBytes, mtimeMs int64, lastLine int, nowMs int64) {
	tx.Exec(`INSERT OR REPLACE INTO chat_index_watermarks (source_file, size_bytes, mtime_ms, last_line, last_indexed_at_ms) VALUES (?, ?, ?, ?, ?)`,
		sourceFile, sizeBytes, mtimeMs, lastLine, nowMs)
}

func buildStableKey(record JsonlRecord, sourceFile string, sourceLine int, rawJson string) string {
	if record.ID != "" {
		return "msg:" + record.ID
	}
	if record.Message != nil && record.Message.ID != "" {
		return "msg:" + record.Message.ID
	}
	// Fallback: source_file + source_line + hash.
	hash := sha256.Sum256([]byte(rawJson))
	return fmt.Sprintf("line:%s:%d:%x", sourceFile, sourceLine, hash[:8])
}

func extractText(content interface{}) string {
	if content == nil {
		return ""
	}
	if s, ok := content.(string); ok {
		return strings.TrimSpace(s)
	}
	blocks, ok := content.([]interface{})
	if !ok {
		return ""
	}
	var parts []string
	for _, block := range blocks {
		m, ok := block.(map[string]interface{})
		if !ok {
			continue
		}
		t, _ := m["type"].(string)
		if t != "text" {
			continue
		}
		text, _ := m["text"].(string)
		if text == "" {
			continue
		}
		// Skip large base64 payloads.
		if len(text) > 10000 && strings.HasPrefix(text, "data:") {
			parts = append(parts, "[media]")
			continue
		}
		normalized := strings.TrimSpace(text)
		if normalized != "" {
			parts = append(parts, normalized)
		}
	}
	return strings.Join(parts, " ")
}

func extractSessionID(record JsonlRecord, sourceFile string) string {
	if record.ID != "" {
		return record.ID
	}
	// Try to extract from file path (e.g., .../sessions/abc123.jsonl).
	base := strings.TrimSuffix(sourceFile, ".jsonl")
	idx := strings.LastIndex(base, "/")
	if idx < 0 {
		idx = strings.LastIndex(base, "\\")
	}
	if idx >= 0 {
		return base[idx+1:]
	}
	return "unknown"
}

func extractChannelFromSessionKey(sessionKey string) string {
	if sessionKey == "" {
		return ""
	}
	parts := strings.Split(sessionKey, ":")
	if len(parts) >= 3 {
		return parts[2]
	}
	return ""
}

File: src/dennou-soul/raw-chat/indexer.ts
import fs from "node:fs";
import { createHash } from "node:crypto";
import type { DatabaseSync } from "node:sqlite";

/**
 * Incremental JSONL indexer for raw chat messages.
 *
 * Reads session JSONL files and indexes new messages into chat_messages + FTS.
 * Uses watermarks to avoid re-scanning the full file on each update.
 *
 * ponytail: Phase 1 — append-oriented, INSERT OR IGNORE, deterministic stable_key.
 */

export type IndexResult = {
  indexed: number;
  skipped: number;
  errors: number;
};

type WatermarkRow = {
  source_file: string;
  size_bytes: number;
  mtime_ms: number;
  last_line: number;
  last_indexed_at_ms: number;
};

type JsonlRecord = {
  type?: string;
  id?: string;
  message?: {
    role?: string;
    content?: unknown;
    id?: string;
    timestamp?: number;
    idempotencyKey?: string;
    [key: string]: unknown;
  };
  timestamp?: number;
  [key: string]: unknown;
};

/**
 * Build a deterministic stable_key for a JSONL message record.
 * Priority: message.id → source_file + source_line + hash(raw_json).
 */
function buildStableKey(
  record: JsonlRecord,
  sourceFile: string,
  sourceLine: number,
  rawJson: string,
): string {
  const messageId = record.id || record.message?.id;
  if (messageId && typeof messageId === "string") {
    return `msg:${messageId}`;
  }
  const hash = createHash("sha256").update(rawJson).digest("hex").slice(0, 16);
  return `line:${sourceFile}:${sourceLine}:${hash}`;
}

/**
 * Extract searchable text from a JSONL message record's content field.
 * Strips large base64 payloads and media metadata.
 */
function extractText(content: unknown): string | null {
  if (typeof content === "string") {
    const trimmed = content.replace(/\s+/g, " ").trim();
    return trimmed || null;
  }
  if (!Array.isArray(content)) {
    return null;
  }
  const parts: string[] = [];
  for (const block of content) {
    if (!block || typeof block !== "object") {
      continue;
    }
    const record = block as { type?: unknown; text?: unknown };
    if (record.type !== "text" || typeof record.text !== "string") {
      continue;
    }
    // Skip large base64 payloads (media inline)
    if (record.text.length > 10_000 && /^data:[a-z]+\/[a-z]+;base64,/.test(record.text)) {
      parts.push("[media]");
      continue;
    }
    const normalized = record.text.replace(/\s+/g, " ").trim();
    if (normalized) {
      parts.push(normalized);
    }
  }
  return parts.length > 0 ? parts.join(" ") : null;
}

/**
 * Index a single session JSONL file incrementally.
 */
export function indexSessionFile(
  db: DatabaseSync,
  sourceFile: string,
  agentId: string,
  sessionKey?: string,
): IndexResult {
  const result: IndexResult = { indexed: 0, skipped: 0, errors: 0 };

  if (!fs.existsSync(sourceFile)) {
    return result;
  }

  let stat: fs.Stats;
  try {
    stat = fs.statSync(sourceFile);
  } catch {
    return result;
  }

  // Check watermark
  const watermark = getWatermark(db, sourceFile);
  const lastLine = watermark?.last_line ?? 0;

  // If file hasn't changed, skip
  if (
    watermark &&
    watermark.size_bytes === stat.size &&
    watermark.mtime_ms === stat.mtimeMs
  ) {
    return result;
  }

  let raw: string;
  try {
    raw = fs.readFileSync(sourceFile, "utf-8");
  } catch {
    return result;
  }

  const lines = raw.split(/\r?\n/);
  const now = Date.now();

  const insertMsg = db.prepare(`
    INSERT OR IGNORE INTO chat_messages (
      stable_key, session_id, session_key, agent_id, channel,
      message_id, parent_id, role, timestamp_ms, timestamp_iso,
      date_key, text, raw_json, source_file, source_line, indexed_at_ms
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `);

  // Use manual transaction for atomicity (node:sqlite DatabaseSync doesn't have .transaction())
  try {
    db.exec(`BEGIN IMMEDIATE`);
    for (let i = lastLine; i < lines.length; i++) {
      const line = lines[i];
      if (!line?.trim()) {
        continue;
      }

      let record: JsonlRecord;
      try {
        record = JSON.parse(line);
      } catch {
        result.errors++;
        continue;
      }

      // Only index message records
      if (!record || record.type !== "message" || !record.message) {
        continue;
      }

      const msg = record.message;
      const role = typeof msg.role === "string" ? msg.role : null;
      if (!role) {
        result.errors++;
        continue;
      }

      const text = extractText(msg.content);
      if (!text) {
        result.skipped++;
        continue;
      }

      const stableKey = buildStableKey(record, sourceFile, i + 1, line);
      const timestampMs = typeof msg.timestamp === "number"
        ? msg.timestamp
        : typeof record.timestamp === "number"
          ? record.timestamp
          : now;
      const timestampIso = new Date(timestampMs).toISOString();
      const dateKey = timestampIso.slice(0, 10);

      // Parse sessionKey components if available
      const channel = sessionKey ? extractChannelFromSessionKey(sessionKey) : null;
      const messageId = typeof msg.id === "string" ? msg.id : null;
      const parentId = typeof msg.parentId === "string" ? msg.parentId : null;

      // Extract session_id from the JSONL header or file path
      const sessionId = extractSessionId(record, sourceFile);

      try {
        insertMsg.run(
          stableKey,
          sessionId,
          sessionKey ?? null,
          agentId,
          channel,
          messageId,
          parentId,
          role,
          timestampMs,
          timestampIso,
          dateKey,
          text,
          line,
          sourceFile,
          i + 1,
          now,
        );

        // Insert into FTS if the row was actually inserted (not ignored)
        const rowId = db.prepare(`SELECT last_insert_rowid() as id`).get() as { id: number };
        if (rowId.id > 0) {
          db.prepare(`
            INSERT INTO chat_messages_fts (rowid, text, role, session_id, timestamp_iso)
            VALUES (?, ?, ?, ?, ?)
          `).run(rowId.id, text, role, sessionId, timestampIso);
          result.indexed++;
        } else {
          result.skipped++;
        }
      } catch {
        result.errors++;
      }
    }

    // Update watermark
    updateWatermark(db, sourceFile, stat.size, stat.mtimeMs, lines.length, now);
    db.exec(`COMMIT`);
  } catch {
    try {
      db.exec(`ROLLBACK`);
    } catch {
      // Rollback also failed; best-effort.
    }
  }

  return result;
}

function getWatermark(db: DatabaseSync, sourceFile: string): WatermarkRow | null {
  try {
    return db.prepare(`SELECT * FROM chat_index_watermarks WHERE source_file = ?`).get(
      sourceFile,
    ) as WatermarkRow | undefined ?? null;
  } catch {
    return null;
  }
}

function updateWatermark(
  db: DatabaseSync,
  sourceFile: string,
  sizeBytes: number,
  mtimeMs: number,
  lastLine: number,
  nowMs: number,
): void {
  db.prepare(`
    INSERT OR REPLACE INTO chat_index_watermarks
    (source_file, size_bytes, mtime_ms, last_line, last_indexed_at_ms)
    VALUES (?, ?, ?, ?, ?)
  `).run(sourceFile, sizeBytes, mtimeMs, lastLine, nowMs);
}

function extractSessionId(record: JsonlRecord, sourceFile: string): string {
  // Try to get session ID from the JSONL header
  if (record.id && typeof record.id === "string") {
    return record.id;
  }
  // Fall back to extracting from file path (e.g., .../sessions/abc123.jsonl)
  const match = sourceFile.match(/([a-z0-9][\w.-]*?)\.jsonl$/i);
  return match?.[1] ?? "unknown";
}

function extractChannelFromSessionKey(sessionKey: string): string | null {
  // Session keys are typically: agent:<agentId>:<channel>:<type>:<id>
  const parts = sessionKey.split(":");
  if (parts.length >= 3) {
    return parts[2] ?? null;
  }
  return null;
}
